### PR TITLE
Add Nicomanga

### LINKED ISSUE #14: NicoManga
### Source name

NicoManga

### Source link

https://nicomanga.com/

### Source language

Japanese - 日本語

### Other details

image

### Acknowledgements

- [X] I have checked that the extension does not already exist by searching the [GitHub repository](https://github.com/keiyoushi/extensions-source/) and verified it does not appear in the code base.
- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a meaningful title with the source name.
- [X] I will fill out all of the requested information in this form.

### PR DESCRIPTION
**Pull request by CodeSpoof**
Checklist:

Closes #14 

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [ ] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [x] Have not changed source names
- [x] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio

### CODE REVIEW

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    override fun latestUpdatesRequest(page: Int): Request = GET("$baseUrl/manga-list.html?page=$page&sort=last_update&sort_type=DESC")
+
+    override fun popularMangaParse(response: Response): MangasPage = mangaListParse(response)
+
+    override fun popularMangaRequest(page: Int): Request = GET("$baseUrl/manga-list.html?page=$page&sort=views&sort_type=DESC")
+
+    override fun searchMangaParse(response: Response): MangasPage = mangaListParse(response)
+
+    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request = GET("$baseUrl/manga-list.html?page=$page&artist=&author=&group=&m_status=&name=$query&genre=&ungenre=&magazine=&sort=last_update&sort_type=DESC")
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:74]: Use HttpUrl.Builder
```suggestion
    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request {
        val url = "$baseUrl/manga-list.html".toHttpUrl().newBuilder()
            .addQueryParameter("page", page.toString())
            .addQueryParameter("artist", "")
            .addQueryParameter("author", "")
            .addQueryParameter("group", "")
            .addQueryParameter("m_status", "")
            .addQueryParameter("name", query)
            .addQueryParameter("genre", "")
            .addQueryParameter("ungenre", "")
            .addQueryParameter("magazine", "")
            .addQueryParameter("sort", "last_update")
            .addQueryParameter("sort_type", "DESC")
            .build()

        return GET(url, headers)
    }
```

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+        if ((request.headers["Referer"] ?: "") != "") {
+            headers.add("Referer", baseUrl)
+        }
+        chain.proceed(request.newBuilder().headers(headers.build()).build())
+    }.build()
+
+    private fun mangaListParse(response: Response): MangasPage {
+        val doc = Jsoup.parse(response.body.string())
+        val mangaList: ArrayList<Element> = doc.select(".row > .thumb-item-flow")
+        val hasNextPage =
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:42]: why not use `ParsedHttpSource`?
You can use something like `.pagination > li:last-child > a:not(.disabled)` for `*MangaNextPageSelector`

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        return chapters
+    }
+
+    override fun chapterListRequest(manga: SManga): Request {
+        val slug = urlRegex.find(manga.url)?.groupValues?.get(0) ?: ""
+        return GET("$baseUrl/app/manga/controllers/cont.Listchapterapi.php?slug=$slug")
+    }
+
+    override fun pageListParse(response: Response): List<Page> {
+        val doc = Jsoup.parse(response.body.string())
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:127]: this is usually an indication you should use `ParsedHttpSource`, so you can override `pageListParse(document: Document)` etc.

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+        val pages = ArrayList<Page>()
+        // Nicovideo will refuse to serve any pages if the user has not logged in
+        if (!doc.select("#login_manga").isEmpty()) {
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:130]:
```suggestion
        if (doc.select("#login_manga").isNotEmpty()) {
```

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+            throw SecurityException("Not logged in. Please login via WebView first")
+        }
+        val pageList = doc.select("#page_contents > li")
+        for (page in pageList) {
+            val pageNumber = page.attr("data-page-index").toInt()
+            val url = page.select("div > img").attr("data-original")
+            pages.add(Page(pageNumber, url, url))
+        }
+        return pages
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:139]:
```suggestion
        return doc.select("#page_contents > li").map { page ->
            val pageNumber = page.attr("data-page-index").toInt()
            val url = page.select("div > img").attr("data-original")
            Page(pageNumber, url, url)
        }
```

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+                },
+            )
+        }
+        editor.apply()
+        return chapters
+    }
+
+    override fun chapterListRequest(manga: SManga): Request {
+        val slug = urlRegex.find(manga.url)?.groupValues?.get(0) ?: ""
+        return GET("$baseUrl/app/manga/controllers/cont.Listchapterapi.php?slug=$slug")
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:123]: if slug can contain special characters, see if you need HttpUrl.Builder

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    private val floatRegex: Regex = "\\d+(?:\\.\\d+)?".toRegex()
+    private val chapterIdRegex: Regex = "(?<=imgsListchap\\()\\d+".toRegex()
+    override val client: OkHttpClient = network.cloudflareClient.newBuilder().addInterceptor { chain ->
+        val request = chain.request()
+        val headers = request.headers.newBuilder()
+        if ((request.headers["Referer"] ?: "") != "") {
+            headers.add("Referer", baseUrl)
+        }
+        chain.proceed(request.newBuilder().headers(headers.build()).build())
+    }.build()
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:37]: Is there a reason you can't override `headersBuilder()`?
```kt
    override fun headersBuilder() = super.headersBuilder()
        .add("Referer", "$baseUrl/")
```

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        val hasNextPage =
+            if (doc.select(".pagination li:last-of-type").size > 0 &&
+                doc.select(".pagination li:last-of-type")[0].text() == "»"
+            ) {
+                doc.select(".pagination li:last-of-type a.disabled").size == 0
+            } else {
+                doc.select(".pagination li:last-of-type a.active").size == 0
+            }
+        val mangas = ArrayList<SManga>()
+        for (manga in mangaList) {
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:51]: You can use a `map` here instead

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+            val url = manga.selectFirst(".series-title a")?.attr("href") ?: ""
+            mangas.add(
+                SManga.create().apply {
+                    setUrlWithoutDomain(URL(URL(baseUrl), url).toString())
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:55]: Is this really necessary?

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+    override fun chapterListParse(response: Response): List<SChapter> {
+        val doc = Jsoup.parse(response.body.string())
+        val chapters = ArrayList<SChapter>()
+        val chapterList = doc.select("ul > a")
+        val mangaId = response.request.url.toUrl().toString().substringAfterLast('/').substringBeforeLast('.')
+        val sharedPref = application.getSharedPreferences("source_${id}_time_found:$mangaId", 0)
+        val editor = sharedPref.edit()
+        var lastNum = 0f
+        for (chapter in chapterList) {
```
Comment (vetleledaal) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:102]: You can use `.map` here

**Review by vetleledaal — CHANGES_REQUESTED** [review]

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    override val baseUrl: String = "https://nicomanga.com"
+    override val lang: String = "ja"
+    override val name: String = "Nicomanga"
+    override val supportsLatest: Boolean = true
+    private val application: Application by injectLazy()
+    private val thumbnailURLRegex: Regex = "background-image:[^;]url\\s*\\(\\s*'([^?']+)".toRegex()
+    private val statusRegex: Regex = "(?<=-)[^.]+".toRegex()
+    private val urlRegex: Regex = "(?<=manga-)[^/]+(?=\\.html\$)".toRegex()
+    private val floatRegex: Regex = "\\d+(?:\\.\\d+)?".toRegex()
+    private val chapterIdRegex: Regex = "(?<=imgsListchap\\()\\d+".toRegex()
```
Comment (FourTOne5) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:29]: Move these regex inside a companion object

File: src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+                SChapter.create().apply {
+                    name = chapter.attr("title").trim()
+                    setUrlWithoutDomain(URL(URL(baseUrl), chapter.attr("href")).toString())
+                    chapter_number = floatRegex.find(chapter.attr("title").trim())?.groupValues?.get(0)?.toFloat() ?: (lastNum + 0.01f)
+                    lastNum = chapter_number
+                    val dateFound = System.currentTimeMillis()
+                    if (!sharedPref.contains(chapter_number.toString())) {
+                        editor.putLong(chapter_number.toString(), dateFound)
+                    }
+                    date_upload = sharedPref.getLong(chapter_number.toString(), dateFound)
```
Comment (FourTOne5) [inline src/ja/nicomanga/src/eu/kanade/tachiyomi/extension/ja/nicomanga/Nicomanga.kt:113]: If you don't specify date_upload the app will do the same thing so this is unnecessary

**Review by FourTOne5 — CHANGES_REQUESTED** [review]

**Comment by CodeSpoof** [comment]
made an error, had some wrong changes in this commit

**Pull request closed without merging**